Series sorting filtering

thFrgttn opened the pull request:
Because additional series results are now returned due to #143 the series selection window can with a large number of results that are not usually sorted in a useful way.

I've created 3 settings that can help finding the corect series quickly
1. use the publisher black list - can be toggled from the series selction screen, as well as a setting for is default behaviour
2. a setting mak the result initally sortted by start year insead of the default no of issues
3. a setting to initially put exact and near matches at the top of the list

lordwelch committed:
Update travis

hold windows to 3.7.9 as unrar-cffi only has windows wheels for 3.7
switch to using builtin python for macOS
remove ssl dlls from comictagger.spec
require pyinstaller=4.3 to allow macOS codesigning
Update python usage
restrict builds to tags and pull requests

thFrgttn committed:
increase results limit back 100

thFrgttn committed:
added sorting settings - default values true atm

thFrgttn committed:
added sorting and filtering

thFrgttn committed:
improved publisher filtering

thFrgttn committed:
improved publisher filtering

thFrgttn committed:
comment

thFrgttn committed:
fixed sorting crash with missing values

thFrgttn committed:
remove extra coloumn resize

thFrgttn committed:
fixed crash when year had none num chars

thFrgttn committed:
Merge branch 'comictagger:develop' into develop

thFrgttn committed:
added near matches

thFrgttn committed:
added near matches

thFrgttn committed:
fixed bad comit

lordwelch committed:
Improve issue identification

Move title sanitizing code to utils module
Update issue identifier to compare sanitized names

thFrgttn committed:
Merge branch 'comictagger:develop' into develop

thFrgttn committed:
fixed crash converting none to string

thFrgttn committed:
fixed crash filtering publishers if None

thFrgttn committed:
change exact matches to case insensitive

thFrgttn committed:
try blocks for sort/filter - incase of bad data

thFrgttn committed:
try blocks for sort/filter - incase of bad data

thFrgttn committed:
added ui elements for new search settings

thFrgttn committed:
added ui elements for new search settings

thFrgttn committed:
better checkbox label

thFrgttn committed:
saving and loading the sort/filter settings

thFrgttn committed:
changed matching to use utils.sanitized_title

thFrgttn committed:
changed matching to use utils.sanitized_title

lordwelch commented on comictaggerlib/ui/settingswindow.ui:
```diff
@@ -263,6 +263,40 @@
+              </property>
+            </widget>
+            </item>
+              <item>
+               <widget class="QCheckBox" name="cbxUseBlackFilter">
+                <property name="text">
+                 <string>Always use Publisher Blacklist on 'manual' searches</string>
+                </property>
+               </widget>
+              </item>
```
This should be next to the blacklist entry, it makes more sense to have all of the blacklist settings together.
It may make sense to rewrite the Identifier tab to be 'Search' instead.

thFrgttn replied:
I don't have an editor for qt ui files installed, and the layout of Identifier tab is just a bit too complex to edit from editting the xml text, and not something I'm going to be able to do

lordwelch replied:
The editor is qt designer I'll work something in tonight.
Here is a windows build I've used in the past as otherwise you need to install the full Qt buildsystem
https://build-system.fman.io/qt-designer-download

lordwelch replied:
Here are two patches one is just the ui and consistency the other renames blacklist to filter
[0001-Update-the-blacklist-settings-no-rename.patch.txt](https://github.com/comictagger/comictagger/files/7264268/0001-Update-the-blacklist-settings-no-rename.patch.txt)
[0001-Update-the-blacklist-settings-rename-to-filter.patch.txt](https://github.com/comictagger/comictagger/files/7264269/0001-Update-the-blacklist-settings-rename-to-filter.patch.txt)

lordwelch commented on comictaggerlib/ui/volumeselectionwindow.ui:
```diff
@@ -148,6 +148,13 @@
+        <item>
+         <widget class="QCheckBox" name="cbxFilter">
+          <property name="text">
+           <string>Apply Publishers Filter</string>
```
We should be consistent on naming I would probably go with filter

thFrgttn replied:
Changed it to 'Filter Publishers' and added a tooltip mentioning the publisher blackllist

lordwelch commented on comictaggerlib/volumeselectionwindow.py:
```diff
@@ -32,9 +32,11 @@
 from .coverimagewidget import CoverImageWidget
 from comictaggerlib.ui.qtutils import reduceWidgetFontSize, centerWindowOnParent
+
+from comictaggerlib import settings
```
This import isn't used

thFrgttn replied:
Your correct, settings are imported 4 lines above

lordwelch commented on comictaggerlib/volumeselectionwindow.py:
```diff
@@ -124,13 +126,18 @@ def __init__(self, parent, series_name, issue_number, year, issue_count,
         self.cover_index_list = cover_index_list
         self.cv_search_results = None
 
+        self.use_blackList = self.settings.use_publisher_blacklist_for_manual
```
Calling this always use blacklist/filter is more descriptive and makes it so that you don't need the context of the blacklist help text to understand the usage of the setting

thFrgttn replied:
agreed

lordwelch commented on comictaggerlib/volumeselectionwindow.py:
```diff
@@ -334,6 +345,40 @@ def searchComplete(self):
+        # filter the blacklisted publishers if setting set
+        if self.use_blackList:
+            try:
+                publisher_blacklist = set([s.strip().lower() for s in self.settings.id_publisher_blacklist.split(',')])
```
So I just discovered that there are [set comprehensions](https://dmerej.info/blog/post/hello-flake8/) you just use curly braces instead of straight braces.
```suggestion
                publisher_blacklist = {s.strip().lower() for s in self.settings.id_publisher_blacklist.split(',')}
```

thFrgttn replied:
This now fails if the start_year has a non numerical value -
Comic VInes has a 'Captain Marvel' series with a start year of '1976?'

lordwelch replied:
What causes the failure this is creating a set for the publisher filter, where does the year come into play?

thFrgttn replied:
Think I reported this wrong, I must have tested bad code
(need to stop dong things late a night)

lordwelch commented on comictaggerlib/volumeselectionwindow.py:
```diff
@@ -334,6 +345,40 @@ def searchComplete(self):
+            try:
+                publisher_blacklist = set([s.strip().lower() for s in self.settings.id_publisher_blacklist.split(',')])
+                # use '' as publisher name if None
+                self.cv_search_results = list(filter(lambda d: ('' if d['publisher'] is None else str(d['publisher']['name']).lower()) not in publisher_blacklist, self.cv_search_results))
```
There's nothing wrong with this but I count 8 different things going on in this statement and maintaining a loop constructing a new list is much simpler

thFrgttn replied:
this altered to deal with bad data, which is why it got so busy
my only question about using a loop, is which is faster

lordwelch replied:
Technically it is faster to use a comprehension ([blog](https://towardsdatascience.com/speeding-up-python-code-fast-filtering-and-slow-loops-8e11a09a9c2f) note that all of the code snippets are juypeter notebooks and it is not mentioned that they are not pure python), however the current cap is hardcoded at 500 and even if you remove the cap "batman" only has 1842 series the filter takes 0.0013023210003666463 seconds, the entire searchComplete function takes 1.603541956999834 seconds

If you are going to talk about speed this is not the function to do it at.

When opening a cached search for the first time versus clicking on the filter checkbox to re-run just the sorting and filtering something makes it go faster so the total for the searchComplete function is 0.113114669999959 seconds without the filtering which is still 100 times longer than the filtering takes on 1842 series.

I'll merge it either way but next time someone takes a look at filtering a loop is much easier to parse through in my experience

lordwelch commented on comictaggerlib/volumeselectionwindow.py:
```diff
@@ -334,6 +345,40 @@ def searchComplete(self):
+            except:
+                print('bad data error filtering blacklist publishers')
+
+        # pre sort the data - so that we can put exact matches first afterwards
+        # compare as str incase extra chars ie. '1976?'
+        # - missing (none) values being converted to 'None' - consistant with prior behaviour in v1.2.3
+        # sort by start_year if set
+        if self.settings.sort_series_by_year:
+            try:
+                self.cv_search_results = sorted(self.cv_search_results, key = lambda i: (str(i['start_year']), str(i['count_of_issues'])), reverse=True)
+            except:
+                print('bad data error sorting results by start_year,count_of_issues')
+        else:
+            try:
+                self.cv_search_results = sorted(self.cv_search_results, key = lambda i: str(i['count_of_issues']), reverse=True)
+            except:
+                print('bad data error sorting results by count_of_issues')
```
natsort handles different types well also operator.itemgetter makes simple keys much easier

```suggestion
        # pre sort the data, so that we can put exact matches first
        # sort by start_year if set
        if self.settings.sort_series_by_year:
            try:
                self.cv_search_results = natsorted(self.cv_search_results, key=itemgetter('start_year', 'count_of_issues'), reverse=True)
            except:
                print('error sorting results by start_year,count_of_issues')
        else:
            try:
                self.cv_search_results = natsorted(self.cv_search_results, key=itemgetter('count_of_issues'), reverse=True)
            except:
                print('error sorting results by count_of_issues')
```

thFrgttn replied:
This fails sorting records containing None values
when you search Comic Vine for a series called 'Excalibur', one of the results is a bed result that just has None in the other fields

lordwelch replied:
This is true I only tested when comparing strings and None which are comparable however numeric types are not comparable with None

lordwelch commented on comictaggerlib/volumeselectionwindow.py:
```diff
@@ -375,9 +420,9 @@ def searchComplete(self):
             row += 1
 
-        self.twList.resizeColumnsToContents()
+        # redundant
+        #self.twList.resizeColumnsToContents()
```
```suggestion
```

lordwelch reviewed (changes requested):
Hello, I put a few notes in on what I noticed, also could you squash your commits it makes for a much cleaner git history.

lordwelch committed:
Add GitHub Actions yaml file (#201)

Upload artifacts this allows easy testing of macOS and Windows binaries
Update unrar-cffi for Python 3.9 wheels

thFrgttn committed:
Update comictaggerlib/volumeselectionwindow.py

removing redundant line

Co-authored-by: Timmy Welch <timmy@narnian.us>

thFrgttn committed:
changed setting to always_use_publisher_blacklist

removing redundant line

Co-authored-by: Timmy Welch <timmy@narnian.us>

Co-authored-by: Timmy Welch <timmy@narnian.us>

lordwelch commented:
Also binaries built using pyinstaller are uploaded for macos and windows if you look under the checks

thFrgttn committed:
Update comictaggerlib/volumeselectionwindow.py

Co-authored-by: Timmy Welch <timmy@narnian.us>

thFrgttn committed:
Merge branch 'comictagger:develop' into seriesSortingFiltering

thFrgttn committed:
Merge branch seriesSortingFiltering

lordwelch commented:
@thFrgttn If you squash your commits I will merge this pull request
